### PR TITLE
Revise generators

### PR DESCRIPTION
**Pull request by nkx111**
Large nkx111  [](https://gitlab.cern.ch/rest-for-physics/restg4/-/commits/revise_generators) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=rest-for-physics&utm_campaign=badge_info)

Revise generators class structure, splitted generator type and shape definition leads to different interface for restG4. Position generation is also splitted to multiple methods: GenPositionOnGDMLVolume, GenPositionOnBoxSurface, etc.

### CODE REVIEW

**Comment by nkx111** [comment]
There is a conflict that, the method `SetParticlePosition(n)` is removed which was used for individual particle position generation. In my opinion, in one event, all particles should be origined from one single vertex.

**Comment by jgalan** [comment]
> There is a conflict that, the method `SetParticlePosition(n)` is removed which was used for individual particle position generation. In my opinion, in one event, all particles should be origined from one single vertex, so I leave only `SetParticlePosition()` to set the position once.

Ok, it makes sense, but it seems there are still other conflicts?

**Comment by jgalan** [comment]
Muon example is failing https://gitlab.cern.ch/rest-for-physics/restg4/-/jobs/14026972

**Comment by nkx111** [comment]
OK, I have fixed those pipelines. 

 > Ok, it makes sense, but it seems there are still other conflicts?

There is another conflict about direction generation. I think the current angular distribution is too complex with so many special cases. I just annotate some of them. This may have some influence on biasing result.

**Review by jgalan — APPROVED** [review]

**Comment by jgalan** [comment]
Ok, thats fine for me, I guess we still need to resolve some conflicts before we are allowed to merge. I imagine there were some commits at master that were not at this branch?

File: src/PrimaryGeneratorAction.cc [hunk 1]
```diff
@@ -17,20 +17,22 @@
...
 
+double GeneratorRndm() { return G4UniformRand(); }
+
 PrimaryGeneratorAction::PrimaryGeneratorAction(DetectorConstruction* pDetector)
     : G4VUserPrimaryGeneratorAction(), fParticleGun(0), fDetector(pDetector) {
     G4int n_particle = 1;
     fParticleGun = new G4ParticleGun(n_particle);
 
-    nCollections = restG4Metadata->GetPrimaryGenerator().GetNumberOfCollections();
+    fGeneratorSpatialDensityFunction = NULL;
```
Comment (juanangp) [inline src/PrimaryGeneratorAction.cc:31]: `nullptr`

File: src/PrimaryGeneratorAction.cc [hunk 2]
```diff
@@ -39,132 +41,134 @@ PrimaryGeneratorAction::~PrimaryGeneratorAction() { delete fParticleGun; }
...
+    }
 
-        throw "Invalid generator type";
+    fSpectrumIntegral = fSpectrum->Integral(startEnergyBin, endEnergyBin);
+}
+
+void PrimaryGeneratorAction::SetGeneratorSpatialDensity(TString str) {
+    string expression = (string)str;
+    if (fGeneratorSpatialDensityFunction != NULL) delete fGeneratorSpatialDensityFunction;
+    if (expression.find_first_of("xyz") == -1) {
```
Comment (juanangp) [inline src/PrimaryGeneratorAction.cc:85]: `expression.find_first_of("xyz") == std::string::npos)`

File: src/PrimaryGeneratorAction.cc [hunk 3]
```diff
@@ -39,132 +41,134 @@ PrimaryGeneratorAction::~PrimaryGeneratorAction() { delete fParticleGun; }
...
+        fGeneratorSpatialDensityFunction = NULL;
```
Comment (juanangp) [inline src/PrimaryGeneratorAction.cc:86]: `nullptr`

File: src/PrimaryGeneratorAction.cc [hunk 4]
```diff
@@ -39,132 +41,134 @@ PrimaryGeneratorAction::~PrimaryGeneratorAction() { delete fParticleGun; }
...
-        fParticleGun->GeneratePrimaryVertex(geant4_event);
+            fParticleGun->GeneratePrimaryVertex(geant4_event);
+        }
     }
 }
 
 //_____________________________________________________________________________
-G4ParticleDefinition* PrimaryGeneratorAction::SetParticleDefinition(int n) {
-    string particle_name = (string)restG4Metadata->GetParticleSource(n).GetParticleName();
+G4ParticleDefinition* PrimaryGeneratorAction::SetParticleDefinition(Int_t n, TRestGeant4Particle p) {
```
Comment (juanangp) [inline src/PrimaryGeneratorAction.cc:130]: Would be better to pass TRestGeant4Particle as reference

`PrimaryGeneratorAction::SetParticleDefinition(Int_t n, const TRestGeant4Particle &p)`

File: src/PrimaryGeneratorAction.cc [hunk 5]
```diff
@@ -39,132 +41,134 @@ PrimaryGeneratorAction::~PrimaryGeneratorAction() { delete fParticleGun; }
...
+    fParticleGun->SetParticleDefinition(particle);
 
     restG4Event->SetPrimaryEventParticleName(particle_name);
 
-    return fParticle;
+    return particle;
 }
 
-void PrimaryGeneratorAction::SetParticleDirection(int n) {
+void PrimaryGeneratorAction::SetParticleDirection(Int_t n, TRestGeant4Particle p) {
```
Comment (juanangp) [inline src/PrimaryGeneratorAction.cc:166]: Same pass TRestGeant4Particle as reference

File: src/PrimaryGeneratorAction.cc [hunk 6]
```diff
@@ -366,10 +369,10 @@ void PrimaryGeneratorAction::SetParticleDirection(int n) {
 }
 
 //_____________________________________________________________________________
-void PrimaryGeneratorAction::SetParticleEnergy(int n) {
+void PrimaryGeneratorAction::SetParticleEnergy(Int_t n, TRestGeant4Particle p) {
```
Comment (juanangp) [inline src/PrimaryGeneratorAction.cc:372]: Same, pass TRestGeant4Particle as reference

File: src/PrimaryGeneratorAction.cc [hunk 7]
```diff
@@ -452,192 +455,63 @@ void PrimaryGeneratorAction::SetParticlePosition() {
...
-        x = x + center.X();
-        y = y + center.Y();
-        z = z + center.Z();
-    } else {
-        G4cout << "WARNING! Generator type was not recognized. Launching particle "
-                  "from origin (0,0,0)"
-               << G4endl;
+        // use the density funciton. If the density is small, then val2 is small, we are more
+        // likely to regenerate the particle position
+        if (fGeneratorSpatialDensityFunction != NULL) {
```
Comment (juanangp) [inline src/PrimaryGeneratorAction.cc:499]: `if(fGeneratorSpatialDensityFunction != nullptr)` or 'if (fGeneratorSpatialDensityFunction)`

File: src/PrimaryGeneratorAction.cc [hunk 8]
```diff
@@ -452,192 +455,63 @@ void PrimaryGeneratorAction::SetParticlePosition() {
...
+            if (val2 > 1) {
+                cout << "error! Generator density function > 1 at position (" << x << ", " << y << ", " << z
+                     << "), check your definition!" << endl;
+                abort();
+            }
+            if (val1 > val2) {
+                continue;
+            }
+        }
+        break;
```
Comment (juanangp) [inline src/PrimaryGeneratorAction.cc:511]: Please, consider to move to:
```do {

} while (val1 > val2);```

Code will be more readable.

**Review by lobis — APPROVED** [review]

**Pull request merged**